File: src/ArtificialViscosity/MorrisMonaghanReducingViscosity.cc
//---------------------------------Spheral++----------------------------------//
// A simple form of the reducing artificial viscosity from Morris & Monaghan.
//----------------------------------------------------------------------------//
#include "FileIO/FileIO.hh"
#include "DataOutput/Restart.hh"
#include "Field/FieldList.hh"
#include "DataBase/DataBase.hh"
#include "DataBase/State.hh"
#include "DataBase/StateDerivatives.hh"
#include "NodeList/FluidNodeList.hh"
#include "Neighbor/Neighbor.hh"
#include "Material/EquationOfState.hh"
#include "Boundary/Boundary.hh"
#include "Hydro/HydroFieldNames.hh"
#include "DataBase/IncrementBoundedState.hh"

#include "MorrisMonaghanReducingViscosity.hh"

using std::vector;
using std::string;
using std::pair;
using std::make_pair;
using std::cout;
using std::cerr;
using std::endl;
using std::min;
using std::max;
using std::abs;

namespace Spheral {
    
//------------------------------------------------------------------------------
// Construct with the given value for the linear and quadratic coefficients.
//------------------------------------------------------------------------------
template<typename Dimension>
MorrisMonaghanReducingViscosity<Dimension>::
MorrisMonaghanReducingViscosity(const Scalar nhQ,
                                const Scalar nhL,
                                const Scalar aMin,
                                const Scalar aMax,
                                const Scalar negligibleSoundSpeed):
  Physics<Dimension>(),
  mnhQ(nhQ),
  mnhL(nhL),
  maMin(aMin),
  maMax(aMax),
  mNegCs(negligibleSoundSpeed),
  mClMultiplier(FieldStorageType::CopyFields),
  mCqMultiplier(FieldStorageType::CopyFields),
  mDrvAlphaDtQ(FieldStorageType::CopyFields),
  mDrvAlphaDtL(FieldStorageType::CopyFields),
  mRestart(registerWithRestart(*this)) {
}

//------------------------------------------------------------------------------
// On problem start up, we need to initialize our internal data.
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
initializeProblemStartup(DataBase<Dimension>& dataBase) {
  mClMultiplier = dataBase.newFluidFieldList(0.0, HydroFieldNames::ArtificialViscousClMultiplier);
  mCqMultiplier = dataBase.newFluidFieldList(0.0, HydroFieldNames::ArtificialViscousCqMultiplier);
  mDrvAlphaDtQ = dataBase.newFluidFieldList(0.0, IncrementBoundedState<Dimension, Scalar>::prefix() + HydroFieldNames::ArtificialViscousCqMultiplier);
  mDrvAlphaDtL = dataBase.newFluidFieldList(0.0, IncrementBoundedState<Dimension, Scalar>::prefix() + HydroFieldNames::ArtificialViscousClMultiplier);
}

//------------------------------------------------------------------------------
// Register the state we need/are going to evolve.
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
registerState(DataBase<Dimension>& /*dataBase*/,
              State<Dimension>& state) {
  state.enroll(mCqMultiplier, make_policy<IncrementBoundedState<Dimension, Scalar>>(maMin,maMax));
  state.enroll(mClMultiplier, make_policy<IncrementBoundedState<Dimension, Scalar>>(maMin,maMax));
}
    
//------------------------------------------------------------------------------
// Register the state derivative fields.
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
registerDerivatives(DataBase<Dimension>& /*dataBase*/,
                    StateDerivatives<Dimension>& derivs) {
  derivs.enroll(mDrvAlphaDtQ);
  derivs.enroll(mDrvAlphaDtL);
}

//------------------------------------------------------------------------------
// Boundaries
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
applyGhostBoundaries(State<Dimension>& state,
                     StateDerivatives<Dimension>& derivs) {
  auto ClMult = state.fields(HydroFieldNames::ArtificialViscousClMultiplier, 0.0);
  auto CqMult = state.fields(HydroFieldNames::ArtificialViscousCqMultiplier, 0.0);
  for (auto* bcPtr: range(this->boundaryBegin(), this->boundaryEnd())) {
    bcPtr->applyFieldListGhostBoundary(ClMult);
    bcPtr->applyFieldListGhostBoundary(CqMult);
  }
}
    
//------------------------------------------------------------------------------
// Determine the principle derivatives.
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
evaluateDerivatives(const typename Dimension::Scalar /*time*/,
                    const typename Dimension::Scalar /*dt*/,
                    const DataBase<Dimension>& /*dataBase*/,
                    const State<Dimension>& state,
                    StateDerivatives<Dimension>& derivs) const {
  // Get Qtys for Derivs
  const auto soundSpeed = state.fields(HydroFieldNames::soundSpeed, 0.0);
  const auto massDensity = state.fields(HydroFieldNames::massDensity, 0.0);
  const auto pressure = state.fields(HydroFieldNames::pressure, 0.0);
  const auto Hsmooth = state.fields(HydroFieldNames::H, SymTensor::zero);
  const auto reducingViscosityMultiplierQ = state.fields(HydroFieldNames::ArtificialViscousCqMultiplier, 0.0);
  const auto reducingViscosityMultiplierL = state.fields(HydroFieldNames::ArtificialViscousClMultiplier, 0.0);
    
  // Derivative FieldLists
  const auto DvDx = derivs.fields(HydroFieldNames::velocityGradient, Tensor::zero);
  auto DrvAlphaDtQ = derivs.fields(IncrementBoundedState<Dimension, Scalar>::prefix() + HydroFieldNames::ArtificialViscousCqMultiplier, 0.0);
  auto DrvAlphaDtL = derivs.fields(IncrementBoundedState<Dimension, Scalar>::prefix() + HydroFieldNames::ArtificialViscousClMultiplier, 0.0);
    
  auto numNodeLists = pressure.numFields();
  CHECK(soundSpeed.size() == numNodeLists);
  CHECK(reducingViscosityMultiplierQ.size() == numNodeLists);
  CHECK(reducingViscosityMultiplierL.size() == numNodeLists);
  CHECK(DrvAlphaDtQ.size() == numNodeLists);
  CHECK(DrvAlphaDtL.size() == numNodeLists);
  CHECK(DvDx.size() == numNodeLists);
  CHECK(massDensity.size() == numNodeLists);
  CHECK(Hsmooth.size() == numNodeLists);
        
  // Walk the nodes
  for (auto i = 0u; i < numNodeLists; ++i) {
    const auto numNodes = pressure[i]->numInternalElements();
    for (auto j = 0u; j < numNodes; j++) {
      const Scalar rvQ = reducingViscosityMultiplierQ(i,j);
      const Scalar rvL = reducingViscosityMultiplierL(i,j);
      const Scalar cs = soundSpeed(i,j);
      const Scalar csSafe = (cs*cs + mNegCs*mNegCs)/cs;
      const Scalar pmin = max(mNegCs*mNegCs*massDensity(i,j),abs(pressure(i,j)));
      const Scalar source = max(-DvDx(i,j).Trace(),0.0);
      const Scalar adiabatIndex = max(cs*cs*massDensity(i,j)/pmin,1.0+mNegCs);
      const Scalar decayConstQ = (1.0/mnhQ)*sqrt((adiabatIndex-1.0)/(2.0*adiabatIndex));
      const Scalar decayConstL = (1.0/mnhL)*sqrt((adiabatIndex-1.0)/(2.0*adiabatIndex));
      const Scalar h = 1.0/(Dimension::rootnu(Hsmooth(i,j).Determinant()));
      const Scalar decayTimeQ = h/(decayConstQ*csSafe);
      const Scalar decayTimeL = h/(decayConstL*csSafe);
            
      // safeInverse instead of forcing cs != 0
      DrvAlphaDtQ(i,j) = (maMax-rvQ)*source - (rvQ - maMin)/decayTimeQ;
      DrvAlphaDtL(i,j) = (maMax-rvL)*source - (rvL - maMin)/decayTimeL;
            
      // shock detection switch
    }
  }
}

//------------------------------------------------------------------------------
// Dump the current state to the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
dumpState(FileIO& file, const string& pathName) const {
  file.write(mClMultiplier, pathName + "/ClMultiplier");
  file.write(mCqMultiplier, pathName + "/CqMultiplier");
  file.write(mDrvAlphaDtQ, pathName + "/DrvAlphaDtQ");
  file.write(mDrvAlphaDtL, pathName + "/DrvAlphaDtL");
}
    
//------------------------------------------------------------------------------
// Restore the state from the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
MorrisMonaghanReducingViscosity<Dimension>::
restoreState(const FileIO& file, const string& pathName) {
  file.read(mClMultiplier, pathName + "/ClMultiplier");
  file.read(mCqMultiplier, pathName + "/CqMultiplier");
  file.read(mDrvAlphaDtQ, pathName + "/DrvAlphaDtQ");
  file.read(mDrvAlphaDtL, pathName + "/DrvAlphaDtL");
}
    
//------------------------------------------------------------------------------
// Vote on a time step.
//------------------------------------------------------------------------------
template<typename Dimension>
typename MorrisMonaghanReducingViscosity<Dimension>::TimeStepType
MorrisMonaghanReducingViscosity<Dimension>::
dt(const DataBase<Dimension>& /*dataBase*/,
   const State<Dimension>& /*state*/,
   const StateDerivatives<Dimension>& /*derivs*/,
   const Scalar /*currentTime*/) const {
  return TimeStepType(1.0e100, "Rate of viscosity change -- NO VOTE.");
}
    
}

File: src/PYB11/ArtificialViscosity/MorrisMonaghanReducingViscosity.py
#-------------------------------------------------------------------------------
# MorrisMonaghanReducingViscosity
#-------------------------------------------------------------------------------
from PYB11Generator import *
from Physics import *
from PhysicsAbstractMethods import *
from RestartMethods import *

@PYB11template("Dimension")
class MorrisMonaghanReducingViscosity(Physics):
    """A simple form of the reducing artificial viscosity from Morris & Monaghan.
Computes a correction scaling for the typical Monaghan & Gingold visc.
References:
  Monaghan, J. J, & Gingold, R. A. 1983, J. Comput. Phys., 52, 374
  Monaghan, J. J. 1992, ARA&A, 30, 543
  Morris, J. P., & Monaghan, J. J. 1997, J. Comput. Phys., 136, 41
"""

    PYB11typedefs = """
    using Scalar = typename %(Dimension)s::Scalar;
    using Vector = typename %(Dimension)s::Vector;
    using Tensor = typename %(Dimension)s::Tensor;
    using SymTensor = typename %(Dimension)s::SymTensor;
    using ThirdRankTensor = typename %(Dimension)s::ThirdRankTensor;
    using TimeStepType = typename Physics<%(Dimension)s>::TimeStepType;
    using ResidualType = typename Physics<%(Dimension)s>::ResidualType;
"""

    #...........................................................................
    # Constructors
    def pyinit(self,
               nhQ = ("const Scalar", "5.0"),
               nhL = ("const Scalar", "10.0"),
               aMin = ("const Scalar", "0.1"),
               aMax = ("const Scalar", "2.0"),
               negligibleSoundSpeed = ("const Scalar", "1.0e-10")):
        "Morris-Monaghan reducing viscosity evolution constructor"

    #...........................................................................
    # Methods
    @PYB11virtual
    def applyGhostBoundaries(self,
                             state = "State<%(Dimension)s>&",
                             derivs = "StateDerivatives<%(Dimension)s>&"):
        "Apply boundary conditions to the physics specific fields."
        return "void"

    @PYB11virtual
    def initializeProblemStartup(self,
                                 dataBase = "DataBase<%(Dimension)s>&"):
        "An optional hook to initialize once when the problem is starting up."
        return "void"

    #...........................................................................
    # Properties
    ClMultiplier = PYB11property("const FieldList<%(Dimension)s, Scalar>&", "ClMultiplier",
                                 doc="Correction multiplier for the linear term")
    CqMultiplier = PYB11property("const FieldList<%(Dimension)s, Scalar>&", "CqMultiplier",
                                 doc="Correction multiplier for the quadratic term")
    DrvAlphaDtQ = PYB11property("const FieldList<%(Dimension)s, Scalar>&", "DrvAlphaDtQ")
    DrvAlphaDtL = PYB11property("const FieldList<%(Dimension)s, Scalar>&", "DrvAlphaDtL")
    nhQ = PYB11property("Scalar", "nhQ", "nhQ")
    nhL = PYB11property("Scalar", "nhL", "nhL")
    aMin = PYB11property("Scalar", "aMin", "aMin")
    aMax = PYB11property("Scalar", "aMax", "aMax")
    negligibleSoundSpeed = PYB11property("Scalar", "negligibleSoundSpeed", "negligibleSoundSpeed")

#-------------------------------------------------------------------------------
# Inject abstract interface
#-------------------------------------------------------------------------------
PYB11inject(PhysicsAbstractMethods, MorrisMonaghanReducingViscosity, pure_virtual=False, virtual=True)
PYB11inject(RestartMethods, MorrisMonaghanReducingViscosity)
